### DISCUSSION TITLE
How can I clear/shadow globals before doctests are run?

### DISCUSSION THREAD

**Discussion by adeak** [pytest-dev/pytest#8791] [Q&A] [answered]
The default behaviour of `pytest` is that doctests see the module-global namespace. This seems to correspond to [stdlib `doctest`](https://docs.python.org/3/library/doctest.html#what-s-the-execution-context). I'm trying to isolate doctests from the module global namespace in order to ensure that copy-pasting doctests into a python script or REPL will work as intended. I've found the [`doctest_namespace` fixture](https://docs.pytest.org/en/6.2.x/doctest.html?highlight=execution%20context#doctest-namespace-fixture), but that only seems to _add_ names rather than overwrite them.

Specific example:
```sh
$ tree
.
└── test_package
    └── foo
        ├── bar.py
        └── __init__.py
```

```py
$ cat test_package/foo/bar.py 
val = 42

class Bar:
    def baz(self):
        """Use the quux.

        Examples
        --------

        >>> val == 42
        True

        """
        return True
```

`__init__.py` is empty.

In the above doctest, `val` is pulled in from the module global namespace so the test succeeds:
```
$ pytest -v --doctest-modules test_package/
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.9.1+, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /home/adeak/tmp_env/bin/python3.9
cachedir: .pytest_cache
rootdir: /home/adeak/doctest_test_dir
collected 1 item                                                                                                                                                                                                  

test_package/foo/bar.py::foo.bar.Bar.baz PASSED                                                                                                                                                             [100%]

================================================================================================ 1 passed in 0.17s ================================================================================================
```
I want it to fail (with a `NameError`), because `val` is not defined in the doctest itself.

How can I make the doctest fail, by effectively passing an empty dict to use as the `globals()` of the doctest?

**Comment by RonnyPfannschmidt** [comment, marked as answer]
its not effectively possible

emptying the global name-space practically would break code using the classes

same goes for using modified copies and potentially a number of other details

this should first be experimented with in a plugin to figure how to make it sensible

**Reply by adeak** [reply]
Thanks. Does this also apply if the package being tested is installed? Names would all be created from scratch.

**Reply by adeak** [reply]
The specific use case motivating my question is doctests in PyVista, for instance [here](https://github.com/pyvista/pyvista/blob/274261f684aad2d9f2ce5dcf9c864a7cf2ca2c5a/pyvista/core/filters/poly_data.py#L227-L230):
```py
        >>> import pyvista as pv
        >>> s1 = pv.Sphere()
        >>> s2 = pv.Sphere(center=(0.25, 0, 0))
        >>> intersection, s1_split, s2_split = s1.intersection(s2)
```
There is no need for module-level globals here. And if we accidentally wrote `pyvista` instead of `pv` on the middle lines, the doctest would still pass because we have `pyvista` in the module global namespace.

So I mean to say: when you say this would "break code using the classes", do you mean breaking pytest, or breaking the doctests themselves, or something else?

**Reply by RonnyPfannschmidt** [reply]
it would break doctests themself, a good number of them are not correctly self-contained

**Reply by adeak** [reply]
OK, thanks for the explanation and quick reply :)
